src/Interpreters/Cache/Metadata.cpp

```diff
@@ -249,91 +233,195 @@ LockedKeyPtr CacheMetadata::lockKeyMetadata(
+
+    LOG_DEBUG(log, "Key {} is removed from metadata", key);
+
+    const fs::path key_directory = getPathForKey(key);
+    const fs::path key_prefix_directory = key_directory.parent_path();
+
+    try
+    {
+        if (fs::exists(key_directory))
+            fs::remove_all(key_directory);
+    }
+    catch (...)
+    {
+        LOG_ERROR(log, "Error while removing key {}: {}", key, getCurrentExceptionMessage(true));
+        chassert(false);
+        return next_it;
+    }
+
+    try
+    {
+        std::unique_lock mutex(key_prefix_directory_mutex);
+        if (fs::exists(key_prefix_directory) && fs::is_empty(key_prefix_directory))
+            fs::remove(key_prefix_directory);
+    }
+    catch (...)
+    {
+        LOG_ERROR(log, "Error while removing key {}: {}", key, getCurrentExceptionMessage(true));
+        chassert(false);
+    }
+    return next_it;
+}
 
-        LockedKeyPtr locked_metadata;
+class CleanupQueue
+{
+    friend struct CacheMetadata;
+public:
+    void add(const FileCacheKey & key)
+    {
         {
-            ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheLockKeyMicroseconds);
-            locked_metadata = std::make_unique<LockedKey>(it->second);
+            std::lock_guard lock(mutex);
+            keys.insert(key);
         }
+        CurrentMetrics::add(CurrentMetrics::FilesystemCacheDelayedCleanupElements);
+        cv.notify_one();
+    }
 
-        const auto key_state = locked_metadata->getKeyState();
-        if (key_state == KeyMetadata::KeyState::ACTIVE)
+    void cancel()
+    {
         {
-            /// Key was added back to cache after we submitted it to removal queue.
-            continue;
+            std::lock_guard lock(mutex);
+            cancelled = true;
         }
+        cv.notify_all();
+    }
 
-        chassert(it->second->empty());
-        locked_metadata->markAsRemoved();
-        erase(it);
-        LOG_DEBUG(log, "Key {} is removed from metadata", cleanup_key);
-
-        const fs::path key_directory = getPathForKey(cleanup_key);
-        const fs::path key_prefix_directory = key_directory.parent_path();
+private:
+    std::unordered_set<FileCacheKey> keys;
+    mutable std::mutex mutex;
+    std::condition_variable cv;
+    bool cancelled = false;
+};
 
-        try
-        {
-            if (fs::exists(key_directory))
-                fs::remove_all(key_directory);
-        }
-        catch (...)
+void CacheMetadata::cleanupThreadFunc()
+{
+    while (true)
+    {
+        Key key;
         {
-            LOG_ERROR(log, "Error while removing key {}: {}", cleanup_key, getCurrentExceptionMessage(true));
-            chassert(false);
-            continue;
+            std::unique_lock lock(cleanup_queue->mutex);
+
+            if (cleanup_queue->cancelled)
+                return;
+
+            auto & keys = cleanup_queue->keys;
+            if (keys.empty())
+            {
+                cleanup_queue->cv.wait(lock);
```

Comment:
Probably nitpicking, but I'd add a predicate for spurious wake-ups and check if cancelled right after it (`continue` is not necessary afterward anymore). This allows to avoid unnecessary iterations in case of spurious wake-ups.

```suggestion
                cleanup_queue->cv.wait(lock, [&](){ return cleanup_queue->cancelled || !cleanup_queue->keys->empty(); });
                if (cleanup_queue->cancelled)
                    return;
```